### PR TITLE
New ERT: The (Legally Distinct) Extra-Reality Police Department

### PR DESCRIPTION
**Pull request by Kapu1178**
## About The Pull Request
A new ERT inspired by the Nuclear Throne group, the IDPD of the same name. Sprites by @PestoVerde322.

They are similar to Code Red in strength, or atleast that was the goal.

There are 4 classes, Chief, Gazer, Titan, and Grunt

Chief gets very strong armor and very strong weapons
Titan gets very strong armor, an Energy Barrier Projector, and a sonic jackhammer.
Gazer gets above average armor, medical supplies, a scatter laser, and telekinesis
Grunt gets a combat RCD, Self-Recharging Energy Gun, and below average armor.

## Why It's Good For The Game
Even if it's only used once I will be happy. A new adminbus tool.

## Changelog
:cl: Kapu1178, PestoVerde322
add: Extra-Reality Police Department
imageadd: E.R.P.D hardsuit sprites by PestoVerde322
/:cl:

### CODE REVIEW

**Comment by Crossedfall** [comment]
I don't personally like the name being identical to the source. I do like the new sprites and such. I think they look a lot better and fit in with our current style.  I would strongly suggest extra-dimensional police department, not only because E.D.P.D. is easier to say and it matches up with other threats and references in the code.

**Comment by Kapu1178** [comment]
image

File: code/modules/clothing/spacesuits/miscellaneous.dm [hunk 1]
```diff
@@ -552,3 +553,67 @@ Contains:
...
+	name = "chief suit"
+	desc = "Wearing this makes you feel #1"
+	icon_state = "idpd_chief_suit"
+	helmettype = /obj/item/clothing/head/helmet/space/hardsuit/ert/idpd/chief
+	armor = list("melee" = 80, "bullet" = 70, "laser" = 70, "energy" = 70, "bomb" = 70, "bio" = 100, "rad" = 100, "fire" = 90, "acid" = 90, "stamina" = 90)
+
+/obj/item/clothing/head/helmet/space/hardsuit/ert/idpd/chief
+	name = "chief helmet"
+	desc = "Glorious"
+	icon_state = "idpd_chief_helmet"
```
Comment (PowerfulBacon) [inline code/modules/clothing/spacesuits/miscellaneous.dm:593]: Helmets and hardsuits should probably have the same armor values

**Review by PowerfulBacon — COMMENTED** [review]
The chief seems more powerful than a deathsquad commando, a pulse carbine and a pulse pistol? (Pulse carbine has an insane amount of ammo and basically never needs recharging)

**Comment by WhyIsCaeciliusTaken** [comment]
> The chief seems more powerful than a deathsquad commando, a pulse carbine and a pulse pistol? (Pulse carbine has an insane amount of ammo and basically never needs recharging)

Pulse carbine has 10 shots does it not? It's the pulse rifle that has 200.

File: code/modules/antagonists/ert/idpd.dm [hunk 2]
```diff
@@ -0,0 +1,175 @@
+////Outfit Datums////
+/datum/outfit/ert/idpd
+	name = "I.D.P.D Grunt"
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm]:
```suggestion
	name = "I.D.P.D. Grunt"
```

File: code/modules/antagonists/ert/idpd.dm [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+	if(istype(H.s_store, /obj/item/gun/energy/e_gun))
+		var/obj/item/gun/energy/e_gun/gun = H.s_store
+		gun.selfcharge = TRUE
+
+	var/obj/item/card/id/W = H.wear_id
+	W.icon_state = "centcom"
+	W.access = get_all_accesses()
+	W.access += get_centcom_access("Death Commando")
+	W.assignment = "I.D.P.D Officer"
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm]:
```suggestion
	W.assignment = "I.D.P.D. Officer"
```

Comment (Archanial) [reply code/modules/antagonists/ert/idpd.dm]: there's another few of those missing dots at the end

File: code/modules/antagonists/ert/idpd.dm [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+////IDPD Creation, handled by one_click_antag.dm
+/datum/admins/proc/attemptMakeIDPD(var/list/mob/dead/observer/candidates, var/num_agents, var/datum/ert/ertemplate, var/datum/team/ert/ert_team, var/datum/objective/missionobj) //They don't start at centcom (fuck you linter)
+	if(num_agents > candidates.len || candidates.len == 0)
+		return FALSE
+
+	var/list/mob/dead/observer/chosen = list()
+	for(var/X = 1, X <= num_agents, ++X)
+		var/mob/dead/observer/chosen_one = pick(candidates)
+		candidates -= chosen_one
+		chosen += chosen_one
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm]: there's pick_n_take exactly for that scenario
```suggestion
		chosen += pick_n_take(candidates)
```

File: code/modules/antagonists/ert/idpd.dm [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	priority_announce("FLÄSHYN, ÖHU!", "Higher Dimensional Affairs", ANNOUNCER_SPANOMALIES)
+	playsound(spawnpoint, 'sound/misc/idpd_portal.ogg', 100, 1)
+	addtimer(CALLBACK(src, .proc/spawnIDPD, chosen, ertemplate, idpd_portal, ert_team, missionobj), 20 SECONDS)
+	return TRUE
+
+/datum/admins/proc/spawnIDPD(var/list/mob/dead/observer/chosen, var/datum/ert/ertemplate, var/obj/effect/portal/idpd_portal, var/datum/team/ert/ert_team, var/datum/objective/missionobj)
+	var/spawnloc = idpd_portal.loc
+	while(chosen.len)
+		var/mob/dead/observer/newguy = pick(chosen)
+		chosen -= newguy
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm]:
```suggestion
		var/mob/dead/observer/newguy = pick_n_take(chosen)
```

File: code/modules/antagonists/ert/idpd.dm [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+		if(!newguy.key)
+			continue
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm]: Why would this ever happen?

File: code/modules/antagonists/ert/idpd.dm [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+		//Spawn the body
+		var/mob/living/carbon/human/ERTOperative = new ertemplate.mobtype(spawnloc)
+		newguy.client.prefs.copy_to(ERTOperative)
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm:153]: never assume client is not null

File: code/modules/clothing/spacesuits/miscellaneous.dm (hunk 1 above)
Comment (Archanial) [reply code/modules/clothing/spacesuits/miscellaneous.dm:593]: I agree

File: code/modules/clothing/spacesuits/miscellaneous.dm [hunk 8]
```diff
@@ -552,3 +553,67 @@ Contains:
...
+	name = "gazer helmet"
+	icon_state = "idpd_gazer_helmet"
+	armor = list("melee" = 65, "bullet" = 50, "laser" = 50, "energy" = 50, "bomb" = 50, "bio" = 100, "rad" = 100, "fire" = 80, "acid" = 80, "stamina" = 70)
+
+///Titan
+/obj/item/clothing/suit/space/hardsuit/ert/idpd/titan
+	name = "titan suit"
+	icon_state = "idpd_titan_suit"
+	helmettype = /obj/item/clothing/head/helmet/space/hardsuit/ert/idpd/titan
+	armor = list("melee" = 80, "bullet" = 70, "laser" = 70, "energy" = 70, "bomb" = 90, "bio" = 100, "rad" = 100, "fire" = 90, "acid" = 90, "stamina" = 80)
```
Comment (Archanial) [inline code/modules/clothing/spacesuits/miscellaneous.dm]: isn't that a bit much? It's 2x syndicate hardsuit.

**Review by Archanial — CHANGES_REQUESTED** [review]

File: code/modules/clothing/spacesuits/miscellaneous.dm (hunk 1 above)
Comment (Kapu1178) [reply code/modules/clothing/spacesuits/miscellaneous.dm:593]: accident

File: code/modules/antagonists/ert/idpd.dm (hunk 6 above)
Comment (Kapu1178) [reply code/modules/antagonists/ert/idpd.dm]: Don't ask me. It was taken from E.R.T code.

File: code/modules/clothing/spacesuits/miscellaneous.dm (hunk 8 above)
Comment (Kapu1178) [reply code/modules/clothing/spacesuits/miscellaneous.dm]: Its meant to be a much more durable Code Red hardsuit. The Titan gets one almost on the level of Captain, but doesn't have much offensive presence.

**Comment by Kapu1178** [comment]
> 
> 
> The chief seems more powerful than a deathsquad commando, a pulse carbine and a pulse pistol? (Pulse carbine has an insane amount of ammo and basically never needs recharging)

The Chief is meant to be disproportionately powerful and will likely always be the last one standing.

File: code/modules/clothing/spacesuits/miscellaneous.dm (hunk 8 above)
Comment (Archanial) [reply code/modules/clothing/spacesuits/miscellaneous.dm]: I still think it could be turned down by like 10 points

**Comment by PowerfulBacon** [comment]
The chief alone could kill the entire station though (essentially making the entire squad way too overpowered and in turn worthless since it can't be used)

File: code/modules/clothing/spacesuits/miscellaneous.dm (hunk 8 above)
Comment (Kapu1178) [reply code/modules/clothing/spacesuits/miscellaneous.dm]: adjusted

**Comment by BrumalLake** [comment]
kinda sounds like a chrono legionnaire but i dunno what happened to them nor what idpd are

**Comment by Karmic-Skink** [comment]
Ditto as on NSV. Provide an actual reason why this is good for the game please.

**Comment by Archanial** [comment]
> Ditto as on NSV. Provide an actual reason why this is good for the game please.

Because it gives admins more cool events they can do with a single button click?

**Comment by Karmic-Skink** [comment]
> > Ditto as on NSV. Provide an actual reason why this is good for the game please.
> 
> Because it gives admins more cool events they can do with a single button click?

Define cool.

**Comment by ike709** [comment]
@Karmic-Skink NSV is no longer actively mirroring bee, you can simply not port it or even if you do just _not use it_. We don't need to justify a PR to you.

**Comment by stinky-lizard** [comment]
> Define cool.

Nar-Sie cultists have nearly completely taken over the station. CentCom has caught wind of this, and sends in an emergency response team to quell the problem.

An admin spawning this in might completely change the outcome of the round. While that may seem like it's a bad thing, it means antags have to always stay on their toes - they never know when an ERT might come in. I think keeping antags on their toes is a good thing. Besides, ERTs already exist. This just adds a new one. More variety is good for the game, right? It'll keep the game from getting stale.

**Comment by Karmic-Skink** [comment]
> @Karmic-Skink NSV is no longer actively mirroring bee, you can simply not port it or even if you do just _not use it_. We don't need to justify a PR to you.

The author of the PR seems to disagree.

**Comment by DatBoiTim** [comment]
> kinda sounds like a chrono legionnaire but i dunno what happened to them nor what idpd are

They never get used.

**Comment by PowerfulBacon** [comment]
Fails checks

**Comment by ike709** [comment]
Fails CI

File: code/modules/antagonists/ert/idpd.dm [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+	E.implant(H, null, 1)
+
+	if(istype(H.s_store, /obj/item/gun/energy/e_gun))
+		var/obj/item/gun/energy/e_gun/gun = H.s_store
+		gun.selfcharge = TRUE
+
+	var/obj/item/card/id/W = H.wear_id
+	W.icon_state = "centcom"
+	W.access = get_all_accesses()
+	W.access += get_centcom_access("Death Commando")
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm:40]:
```suggestion
	W.access |= get_centcom_access("Death Commando")
```

File: code/modules/antagonists/ert/idpd.dm [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+	suit_store = /obj/item/gun/energy/laser/scatter
+	l_pocket = /obj/item/reagent_containers/hypospray/combat
+	l_hand = /obj/item/storage/firstaid/tactical
+	r_hand = null
+
+/datum/outfit/ert/erpd/gazer/post_equip(mob/living/carbon/human/H, visualsOnly)
+	. = ..()
+	if(!visualsOnly)
+		H.dna.add_mutation(TK)
+/datum/outfit/ert/erpd/titan
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm:68]:
```suggestion
		H.dna.add_mutation(TK)

/datum/outfit/ert/erpd/titan
```

File: code/modules/antagonists/ert/idpd.dm [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+	if(isliving(mover))
+		return FALSE //None shall pass
+
+////erpd Creation, handled by one_click_antag.dm
+/datum/admins/proc/attemptMakeERPD(var/list/mob/dead/observer/candidates, var/num_agents, var/datum/ert/ertemplate, var/datum/team/ert/ert_team, var/datum/objective/missionobj) //They don't start at centcom (fuck you linter)
+	if(num_agents > candidates.len || candidates.len == 0)
+		return FALSE
+
+	var/list/mob/dead/observer/chosen = list()
+	for(var/X = 1, X <= num_agents, ++X)
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm:127]:
```suggestion
	for(var/X in 1 to num_agents)
```

File: code/modules/antagonists/ert/idpd.dm [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+		chosen_one.orbit(erpd_portal)
+
+	priority_announce("FLÄSHYN, ÖHU!", "Higher Dimensional Affairs", ANNOUNCER_SPANOMALIES)
+	playsound(spawnpoint, 'sound/misc/idpd_portal.ogg', 100, 1)
+	addtimer(CALLBACK(src, .proc/spawnIDPD, chosen, ertemplate, erpd_portal, ert_team, missionobj), 20 SECONDS)
+	return TRUE
+
+/datum/admins/proc/spawnIDPD(var/list/mob/dead/observer/chosen, var/datum/ert/ertemplate, var/obj/effect/portal/erpd_portal, var/datum/team/ert/ert_team, var/datum/objective/missionobj)
+	var/spawnloc = erpd_portal.loc
+	while(chosen.len)
```
Comment (Archanial) [inline code/modules/antagonists/ert/idpd.dm:146]: just do for

**Review by Archanial — CHANGES_REQUESTED** [review]

**Comment by Kapu1178** [comment]
i didnt get notifications for these for some reason. will address.

**Comment by Archanial** [comment]
stale

**Pull request closed without merging**